File: coding-interviewes/coding_interview_150/scala/ProductOfArrayExceptSelf.scala

/* Product Except Self — prefix/suffix. O(n) */
object ProductOfArrayExceptSelf {
  def productExceptSelf(nums:Array[Int]):Array[Int] = {
    val n=nums.length
    val out=Array.fill(n)(1)
    var pref=1; var i=0
    while (i<n){ out(i)=pref; pref*=nums(i); i+=1 }
    var suff=1; i=n-1
    while (i>=0){ out(i)*=suff; suff*=nums(i); i-=1 }
    out
  }
  def main(args:Array[String]):Unit = {
    println(productExceptSelf(Array(1,2,3,4)).mkString("[",",","]"))
  }
}

File: coding-interviewes/coding_interview_150/java/ProductOfArrayExceptSelf.java

/* Product of Array Except Self — prefix & suffix. O(n) time */
import java.util.Arrays;
public class ProductOfArrayExceptSelf {
    public static int[] productExceptSelf(int[] nums){
        int n=nums.length; int[] out=new int[n]; int pref=1;
        for (int i=0;i<n;i++){ out[i]=pref; pref*=nums[i]; }
        int suff=1;
        for (int i=n-1;i>=0;i--){ out[i]*=suff; suff*=nums[i]; }
        return out;
    }
    public static void main(String[] args){
        System.out.println(Arrays.toString(productExceptSelf(new int[]{1,2,3,4})));
    }
}
